File: sdk/python/aragora_sdk/namespaces/webhooks.py
"""
Webhooks Namespace API

Provides methods for webhook management:
- Create and manage webhooks
- Track deliveries
- Test webhook endpoints
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient

_List = list  # Preserve builtin list for type annotations


class WebhooksAPI:
    """
    Synchronous Webhooks API.

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai")
        >>> webhook = client.webhooks.create(
        ...     url="https://example.com/webhook",
        ...     events=["debate.completed", "receipt.created"]
        ... )
        >>> print(webhook["webhook_id"])
    """

    def __init__(self, client: AragoraClient):
        self._client = client

    def list(
        self,
        active_only: bool = True,
        limit: int = 20,
        offset: int = 0,
    ) -> dict[str, Any]:
        """
        List webhooks.

        Args:
            active_only: Only return active webhooks
            limit: Maximum results
            offset: Pagination offset

        Returns:
            List of webhooks
        """
        params: dict[str, Any] = {
            "active_only": active_only,
            "limit": limit,
            "offset": offset,
        }
        return self._client.request("GET", "/api/v1/webhooks", params=params)

    def create(
        self,
        url: str,
        events: _List[str],
        secret: str | None = None,
        description: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """
        Create a webhook.

        Args:
            url: Webhook endpoint URL
            events: List of event types to subscribe to
            secret: Webhook signing secret (auto-generated if not provided)
            description: Webhook description
            headers: Custom headers to include in requests

        Returns:
            Created webhook with secret
        """
        data: dict[str, Any] = {"url": url, "events": events}
        if secret:
            data["secret"] = secret
        if description:
            data["description"] = description
        if headers:
            data["headers"] = headers

        return self._client.request("POST", "/api/v1/webhooks", json=data)

    def get_events(self) -> dict[str, Any]:
        """
        Get available webhook event types.

        Returns:
            List of event types with descriptions
        """
        return self._client.request("GET", "/api/v1/webhooks/events")

    def get_event_categories(self) -> dict[str, Any]:
        """Get webhook event categories."""
        return self._client.request("GET", "/api/v1/webhooks/events/categories")

    def test_slo(self, webhook_id: str | None = None) -> dict[str, Any]:
        """Test webhook SLO delivery."""
        data: dict[str, Any] = {}
        if webhook_id:
            data["webhook_id"] = webhook_id
        return self._client.request("POST", "/api/v1/webhooks/slo/test", json=data)

    def bulk_create(self, webhooks: _List[dict[str, Any]]) -> dict[str, Any]:
        """Create multiple webhooks in bulk."""
        return self._client.request("POST", "/api/v1/webhooks/bulk", json={"webhooks": webhooks})

    def pause_all(self) -> dict[str, Any]:
        """Pause all webhooks."""
        return self._client.request("POST", "/api/v1/webhooks/pause-all")

    def resume_all(self) -> dict[str, Any]:
        """Resume all webhooks."""
        return self._client.request("POST", "/api/v1/webhooks/resume-all")

    def dispatch(self, **kwargs: Any) -> dict[str, Any]:
        """
        Manually dispatch a webhook event.

        Args:
            **kwargs: Event parameters (event_type, payload, webhook_id, etc.)

        Returns:
            Dict with dispatch result and delivery status.
        """
        return self._client.request("POST", "/api/webhooks/dispatch", json=kwargs)

    def list_platforms(self) -> dict[str, Any]:
        """
        List supported webhook platform integrations.

        Returns:
            Dict with supported platforms and their capabilities.
        """
        return self._client.request("GET", "/api/webhooks/platforms")

    # =========================================================================
    # Webhook CRUD
    # =========================================================================

    def get(self, webhook_id: str) -> dict[str, Any]:
        """Get a specific webhook by ID."""
        return self._client.request("GET", f"/api/v1/webhooks/{webhook_id}")

    def update(self, webhook_id: str, **updates: Any) -> dict[str, Any]:
        """Update an existing webhook."""
        return self._client.request("PATCH", f"/api/v1/webhooks/{webhook_id}", json=updates)

    def delete(self, webhook_id: str) -> dict[str, Any]:
        """Delete a webhook."""
        return self._client.request("DELETE", f"/api/v1/webhooks/{webhook_id}")

    def test(self, webhook_id: str) -> dict[str, Any]:
        """Test a webhook by sending a test event."""
        return self._client.request("POST", f"/api/v1/webhooks/{webhook_id}/test")

    # =========================================================================
    # Delivery Management
    # =========================================================================

    def list_dead_letter(self, limit: int | None = None) -> dict[str, Any]:
        """List deliveries in the dead-letter queue."""
        params: dict[str, Any] = {}
        if limit is not None:
            params["limit"] = limit
        return self._client.request("GET", "/api/v1/webhooks/dead-letter", params=params)

    def get_dead_letter(self, dead_letter_id: str) -> dict[str, Any]:
        """Get a specific dead-letter delivery by ID."""
        return self._client.request("GET", f"/api/v1/webhooks/dead-letter/{dead_letter_id}")

    def retry_dead_letter(self, dead_letter_id: str) -> dict[str, Any]:
        """Retry a dead-letter delivery."""
        return self._client.request("POST", f"/api/v1/webhooks/dead-letter/{dead_letter_id}/retry")


class AsyncWebhooksAPI:
    """
    Asynchronous Webhooks API.

    Example:
        >>> async with AragoraAsyncClient(base_url="https://api.aragora.ai") as client:
        ...     webhook = await client.webhooks.create(
        ...         url="https://example.com/webhook",
        ...         events=["debate.completed"]
        ...     )
    """

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    async def list(
        self,
        active_only: bool = True,
        limit: int = 20,
        offset: int = 0,
    ) -> dict[str, Any]:
        """List webhooks."""
        params: dict[str, Any] = {
            "active_only": active_only,
            "limit": limit,
            "offset": offset,
        }
        return await self._client.request("GET", "/api/v1/webhooks", params=params)

    async def create(
        self,
        url: str,
        events: _List[str],
        secret: str | None = None,
        description: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Create a webhook."""
        data: dict[str, Any] = {"url": url, "events": events}
        if secret:
            data["secret"] = secret
        if description:
            data["description"] = description
        if headers:
            data["headers"] = headers

        return await self._client.request("POST", "/api/v1/webhooks", json=data)

    async def get_events(self) -> dict[str, Any]:
        """Get available webhook event types."""
        return await self._client.request("GET", "/api/v1/webhooks/events")

    async def get_event_categories(self) -> dict[str, Any]:
        """Get webhook event categories."""
        return await self._client.request("GET", "/api/v1/webhooks/events/categories")

    async def test_slo(self, webhook_id: str | None = None) -> dict[str, Any]:
        """Test webhook SLO delivery."""
        data: dict[str, Any] = {}
        if webhook_id:
            data["webhook_id"] = webhook_id
        return await self._client.request("POST", "/api/v1/webhooks/slo/test", json=data)

    async def bulk_create(self, webhooks: _List[dict[str, Any]]) -> dict[str, Any]:
        """Create multiple webhooks in bulk."""
        return await self._client.request(
            "POST", "/api/v1/webhooks/bulk", json={"webhooks": webhooks}
        )

    async def pause_all(self) -> dict[str, Any]:
        """Pause all webhooks."""
        return await self._client.request("POST", "/api/v1/webhooks/pause-all")

    async def resume_all(self) -> dict[str, Any]:
        """Resume all webhooks."""
        return await self._client.request("POST", "/api/v1/webhooks/resume-all")

    async def dispatch(self, **kwargs: Any) -> dict[str, Any]:
        """Manually dispatch a webhook event."""
        return await self._client.request("POST", "/api/webhooks/dispatch", json=kwargs)

    async def list_platforms(self) -> dict[str, Any]:
        """List supported webhook platform integrations."""
        return await self._client.request("GET", "/api/webhooks/platforms")

    # =========================================================================
    # Webhook CRUD
    # =========================================================================

    async def get(self, webhook_id: str) -> dict[str, Any]:
        """Get a specific webhook by ID."""
        return await self._client.request("GET", f"/api/v1/webhooks/{webhook_id}")

    async def update(self, webhook_id: str, **updates: Any) -> dict[str, Any]:
        """Update an existing webhook."""
        return await self._client.request("PATCH", f"/api/v1/webhooks/{webhook_id}", json=updates)

    async def delete(self, webhook_id: str) -> dict[str, Any]:
        """Delete a webhook."""
        return await self._client.request("DELETE", f"/api/v1/webhooks/{webhook_id}")

    async def test(self, webhook_id: str) -> dict[str, Any]:
        """Test a webhook by sending a test event."""
        return await self._client.request("POST", f"/api/v1/webhooks/{webhook_id}/test")

    # =========================================================================
    # Delivery Management
    # =========================================================================

    async def list_dead_letter(self, limit: int | None = None) -> dict[str, Any]:
        """List deliveries in the dead-letter queue."""
        params: dict[str, Any] = {}
        if limit is not None:
            params["limit"] = limit
        return await self._client.request("GET", "/api/v1/webhooks/dead-letter", params=params)

    async def get_dead_letter(self, dead_letter_id: str) -> dict[str, Any]:
        """Get a specific dead-letter delivery by ID."""
        return await self._client.request("GET", f"/api/v1/webhooks/dead-letter/{dead_letter_id}")

    async def retry_dead_letter(self, dead_letter_id: str) -> dict[str, Any]:
        """Retry a dead-letter delivery."""
        return await self._client.request(
            "POST", f"/api/v1/webhooks/dead-letter/{dead_letter_id}/retry"
        )

File: sdk/typescript/src/namespaces/webhooks.ts
/**
 * Webhooks Namespace API
 *
 * Provides a namespaced interface for webhook management.
 * Essential for enterprise event-driven integrations.
 */

/**
 * Webhook configuration
 */
export interface Webhook {
  id: string;
  url: string;
  events: string[];
  secret: string;
  active: boolean;
  created_at: string;
  updated_at: string;
  last_triggered_at?: string;
  failure_count: number;
  metadata?: Record<string, unknown>;
}

/**
 * Webhook event type
 */
export interface WebhookEvent {
  name: string;
  description: string;
  category: string;
  payload_schema?: Record<string, unknown>;
}

/**
 * Webhook delivery status
 */
export interface WebhookDelivery {
  id: string;
  webhook_id: string;
  event: string;
  payload: Record<string, unknown>;
  status: 'pending' | 'delivered' | 'failed';
  response_code?: number;
  response_body?: string;
  delivered_at?: string;
  error?: string;
  retry_count: number;
}

/**
 * SLO status response
 */
export interface WebhookSLOStatus {
  delivery_rate: number;
  avg_latency_ms: number;
  p99_latency_ms: number;
  failed_deliveries_24h: number;
  total_deliveries_24h: number;
  status: 'healthy' | 'degraded' | 'critical';
}

/**
 * Create webhook request
 */
export interface CreateWebhookRequest {
  url: string;
  events: string[];
  secret?: string;
  active?: boolean;
  metadata?: Record<string, unknown>;
}

/**
 * Update webhook request
 */
export interface UpdateWebhookRequest {
  url?: string;
  events?: string[];
  secret?: string;
  active?: boolean;
  metadata?: Record<string, unknown>;
  [key: string]: unknown;
}

/**
 * Interface for the internal client used by WebhooksAPI.
 */
interface WebhooksClientInterface {
  get<T>(path: string): Promise<T>;
  post<T>(path: string, body?: unknown): Promise<T>;
  put<T>(path: string, body?: unknown): Promise<T>;
  delete<T>(path: string): Promise<T>;
  request<T = unknown>(method: string, path: string, options?: { params?: Record<string, unknown>; json?: unknown }): Promise<T>;
}

/**
 * Webhooks API namespace.
 *
 * Provides methods for managing webhooks:
 * - Create, update, and delete webhooks
 * - List available webhook events
 * - Test webhook deliveries
 * - Monitor webhook SLO status
 *
 * @example
 * ```typescript
 * const client = createClient({ baseUrl: 'https://api.aragora.ai', apiKey: 'your-key' });
 *
 * // Create a webhook
 * const webhook = await client.webhooks.create({
 *   url: 'https://your-app.com/webhooks',
 *   events: ['debate.completed', 'consensus.reached'],
 * });
 *
 * // List available events
 * const { events } = await client.webhooks.listEvents();
 *
 * // Test a webhook
 * const result = await client.webhooks.test(webhook.id);
 *
 * // Check SLO status
 * const slo = await client.webhooks.getSLOStatus();
 * console.log(`Delivery rate: ${slo.delivery_rate}%`);
 * ```
 */
export class WebhooksAPI {
  constructor(private client: WebhooksClientInterface) {}

  // ===========================================================================
  // Webhook Management
  // ===========================================================================

  /**
   * List all webhooks.
   */
  async list(): Promise<{ webhooks: Webhook[]; total: number }> {
    return this.client.get('/api/v1/webhooks');
  }

  /**
   * Get a specific webhook by ID.
   */
  async get(id: string): Promise<Webhook> {
    return this.client.get(`/api/v1/webhooks/${id}`);
  }

  /**
   * Create a new webhook.
   */
  async create(body: CreateWebhookRequest): Promise<Webhook> {
    return this.client.post('/api/v1/webhooks', body);
  }

  /**
   * Update an existing webhook.
   */
  async update(id: string, body: UpdateWebhookRequest): Promise<Webhook> {
    return this.client.request('PATCH', `/api/v1/webhooks/${id}`, { json: body });
  }

  /**
   * Delete a webhook.
   */
  async delete(id: string): Promise<{ deleted: boolean }> {
    return this.client.delete(`/api/v1/webhooks/${id}`);
  }

  /**
   * Test a webhook by sending a test event.
   */
  async test(id: string): Promise<{ success: boolean; response_code?: number; error?: string }> {
    return this.client.post(`/api/v1/webhooks/${id}/test`);
  }

  // ===========================================================================
  // Events
  // ===========================================================================

  /**
   * List all available webhook events.
   */
  async listEvents(): Promise<{ events: WebhookEvent[] }> {
    return this.client.get('/api/v1/webhooks/events');
  }

  /**
   * List supported webhook platforms.
   * @route GET /api/v1/webhooks/platforms
   */
  async listPlatforms(): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/v1/webhooks/platforms');
  }

  /**
   * Dispatch a webhook event immediately.
   * @route POST /api/v1/webhooks/dispatch
   */
  async dispatch(body: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request('POST', '/api/v1/webhooks/dispatch', { json: body });
  }

  // ===========================================================================
  // SLO Monitoring
  // ===========================================================================

  /**
   * Get webhook SLO status.
   */
  async getSLOStatus(): Promise<WebhookSLOStatus> {
    return this.client.get('/api/v1/webhooks/slo/status');
  }

  /**
   * Test webhook SLO by measuring delivery latency.
   */
  async testSLO(): Promise<{ success: boolean; latency_ms: number }> {
    return this.client.post('/api/v1/webhooks/slo/test');
  }

  // ===========================================================================
  // Event Filtering
  // ===========================================================================

  /**
   * Get event categories.
   */
  async getEventCategories(): Promise<{ categories: { name: string; description: string; events: string[] }[] }> {
    return this.client.request('GET', '/api/v1/webhooks/events/categories');
  }

  // ===========================================================================
  // Bulk Operations
  // ===========================================================================

  /**
   * Pause all webhooks.
   */
  async pauseAll(): Promise<{ paused: number }> {
    return this.client.request('POST', '/api/v1/webhooks/pause-all');
  }

  /**
   * Resume all webhooks.
   */
  async resumeAll(): Promise<{ resumed: number }> {
    return this.client.request('POST', '/api/v1/webhooks/resume-all');
  }

  /**
   * Bulk delete webhooks.
   */
  async bulkDelete(webhookIds: string[]): Promise<{ deleted: number }> {
    return this.client.request('DELETE', '/api/v1/webhooks/bulk', { json: { ids: webhookIds } });
  }

  // ===========================================================================
  // Dead-Letter Queue Management
  // ===========================================================================

  /**
   * List deliveries in the dead-letter queue.
   */
  async listDeadLetter(options?: { limit?: number }): Promise<{ dead_letters: WebhookDelivery[]; count: number }> {
    return this.client.request('GET', '/api/v1/webhooks/dead-letter', { params: options });
  }

  /**
   * Get a specific dead-letter delivery by ID.
   */
  async getDeadLetter(id: string): Promise<WebhookDelivery> {
    return this.client.request('GET', `/api/v1/webhooks/dead-letter/${id}`);
  }

  /**
   * Retry a dead-letter delivery.
   */
  async retryDeadLetter(id: string): Promise<{ retried: boolean; new_delivery_id?: string }> {
    return this.client.request('POST', `/api/v1/webhooks/dead-letter/${id}/retry`);
  }

  /**
   * Delete a dead-letter delivery.
   */
  async deleteDeadLetter(id: string): Promise<{ deleted: boolean }> {
    return this.client.request('DELETE', `/api/v1/webhooks/dead-letter/${id}`);
  }

  /**
   * Get webhook queue statistics.
   */
  async getQueueStats(): Promise<{ stats: { pending: number; processing: number; dead_letter: number; total_delivered: number; total_failed: number } }> {
    return this.client.request('GET', '/api/v1/webhooks/queue/stats');
  }
}
